src/permutat.cc

```diff
@@ -67,6 +69,36 @@
 #include "sysfiles.h"
 #include "trans.h"
 
+}
+
+template<typename T> static inline UInt SIZEBAG_PERM(UInt deg)
+{
+    return sizeof(Obj) + deg * sizeof(T);
+}
+
+template<typename T> static inline Obj NEW_PERM(UInt deg)
+{
+    return NewBag(sizeof(T) == 2 ? T_PERM2 : T_PERM4, SIZEBAG_PERM<T>(deg));
+}
+
+template<typename T> static inline UInt DEG_PERM(Obj perm)
+{
+    return (SIZE_OBJ(perm) - sizeof(Obj)) / sizeof(T);
+}
+
+template<typename T> static inline T * ADDR_PERM(Obj perm)
+{
+    return (T *)(ADDR_OBJ(perm) + 1);
+}
+
+template<typename T> static inline const T * CONST_ADDR_PERM(Obj perm)
+{
+    return (const T *)(CONST_ADDR_OBJ(perm) + 1);
+}
+
+template<typename TL, typename TR> struct ResultType { typedef UInt4 type; };
```

Comment:
Very low-level comment, might want to give this another name before merging because I imagine we'll have other similar types. However, the same thing can be used for perms, trans and pperms.. not sure of a good name.